oeps/oep-0038-Data-Modeling.rst

```diff
@@ -0,0 +1,419 @@
+
+-  Don’t store encoded (pickle, json, other) objects in the database! If
+      you need to run the python environment to decode the data, analyists who use SQL will have a difficult time querying and decoding this data.
+
+Store everything
+~~~~~~~~~~~~~~~~
+
+-  Storage is cheap!
+
+-  If you are unsure whether you should store something in the database or add history the answer is almost always yes. Store the data and add history. It can be removed later if it is found unnecessary.
+
+-  Still not sure? The default answer is yes.
+
+Use methods on models to access data
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- CRUD operations should access models via methods on models (where they exist), instead of querying managers directly.
+
+-  For example, prefer creating something like
+      CourseEnrollment.is_enrolled(...) rather than having views check
+      CourseEnrollment.objects.filter(...).exists().
+
+-  This allows us to more easily make internal representation changes
+      like switching to using a “deleted” flag instead of deleting the
+      row.
+
+-  This also reduces the likelihood that people will query models in a
+      non-performant way (e.g. sorting by an unindexed field).
+
+Enforce logical constraints at the database layer
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+-  Don’t allow impossible states to be represented in the database.
+
+-  If your code expects a 1:1 relationship, use Django's :code:`Unique`_. instead of trying to enforce the constraint in
+      Python.
+.. _Unique: https://docs.djangoproject.com/en/2.2/ref/models/fields/#unique
+
+-  Python will not save you from race conditions. Database constraints will.
+
+-  For example, an enrollment should have a unique constraint on
+      (course_id, user_id), since a given user should only have one
+      enrollment per course. In this case you should use Django's :code:`unique_together`_
+
+Keep indexes to a minimum
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+-  Create indexes only on the fields necessary to make queries performant
+
+-  Keep in mind that indexes cost space and have their own set of performance concerns.
+
+-  Over-indexing data could actually make the database less performant (slower writes/updates)
+
+
+
+Developer Responsibility
+========================
+
+It is the responsibility of the developer to adhere to all of the
+standards in the Data Modeling Standards section of this document.
+
+Code Reviewer Responsibility
+============================
+
+The code reviewer is responsible for ensuring the standards set forth in
+the Data Modeling Standards section of this document are met.
+
+Responsibility for Third-party Service Integrations
+===================================================
+
+Not sure… Adhere to the same standards.
+
+Backward Compatibility
+======================
+
+Data models that are not within the standards of this document should be
```

Comment:
I'm going to change this to no action required. I'm not sure we want to open this can of worms and we would probably be better served by specifically targeting models with issues that will have a material impact on how we are analyzing data or data quality issues that are preventing analysis.